Привет, Никита! Меня зовут Максим Слотин, я буду делать ревью твоего проекта. Давай будем общаться на «ты».

Ты проделал большую исследовательскую работу! Твой проект очень хорошо структурирован, ты делаешь хорошие понятные выводы и выбираешь наглядные типы графиков. Мои комментарии будут в ячейках "Комментарий от ревьюера".

<div class="alert alert-danger"> 
<h2> Комментарии от ревьюера №1</h2>
Красным цветом обозначены комментарии, на которые тебе нужно обратить внимание в первую очередь, чтобы я принял проект. 
</div>
<div class="alert alert-warning">
<h2> Комментарии от ревьюера №1</h2>
Желтым цветом отмечено то, что ты можешь сделать немного по другому. Ты можешь учесть эти комментарии при выполнении следующих проектов.
</div>
<div class="alert alert-success">
<h2> Комментарии от ревьюера №1</h2>
Зелёным цветом обозначены твои удачные решения, которые ты без сомнения можешь применять в будущих проектах.
</div>

#### Пожалуста, не удаляй мои комментарии и относись к ним бережно. При повторной проверке, я буду ориентироваться на них. 

###### Можешь под моими комментариями оставлять свои, в которых опиши свои испарвления. Это облегчит процесс проверки и мне, и тебе  :) 

Например, вот так:

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
</div>
---

</div>
<div class="alert alert-success">
<h2> Комментарии от ревьюера №2</h2>
    
У тебя в итоге вышел отличный проект)
    
Спасибо, что довольно внимательно и терпеливо отреагировал на мои комментарии 😊
    
Я ниже не комментировал конкретных моментов, так как все в целом, что я хотел, ты поправил
    
Желаю тебе дальше успехов в учебе)
</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
import numpy as np
root_df = pd.read_csv('/datasets/data.csv')
root_df.head(30)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

In [2]:
root_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
root_df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

### Вывод

Датасет изучен, получена информация о типе и количестве и типах данных, содержащихся в колонках.

### Шаг 2. Предобработка данных

### Обработка пропусков

#### Поиск null и nan 

In [4]:
root_df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
root_df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

#### Поиск аномалий в столбце children
Идея пришла после того, как на группировке увидел категорию 20 детей, технически это возможно, но нет ни одной семьи в которой было бы больше 5, но меньше 20 детей, так что это скорее ошибка сборка данных. Возможно что при записи в датасет неверно прошла оперция с типом данных и из 2.0 получилось 20. В связи с этим предположением заменим 20 на 2.

<div class="alert alert-success">
<h2> Комментарии от ревьюера №1</h2>
    
Логика супер
</div>

In [6]:
root_df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [7]:
root_df['children'] = root_df['children'].replace(20,2)

#### Поиск негативных значений

In [8]:
#Проверка наличия отрицательных значений total_income
root_df[root_df['total_income'] < 0]['total_income'].count()

0

In [9]:
root_df[root_df['days_employed'] < 0]['days_employed'].count()

15906

In [10]:
root_df[root_df['children'] < 0]['children'].count()

47

In [11]:
def days_employed_to_postive(row):
    return abs(row['days_employed'])

In [12]:
root_df['days_employed'] = root_df.apply(days_employed_to_postive, axis=1)

In [13]:
def children_to_postive(row):
    return abs(row['children'])

In [14]:
root_df['children'] = root_df.apply(children_to_postive, axis=1)

#### Решение проблемы с часами вместо дней 

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    Первое, что нужно сделать - оценить, когда нужно менять значение с часов на дни, а когда оставлять как есть. Для этого определим пороговое значение максимально возможного стажа в днях. 
Возраст в днях в среднем всегда меньше, чем стаж в часах, но при этом всегда больше, чем стаж в днях. Если это условие выполняется, то стаж указан в часах.
</div>

In [15]:
def convert_days_employed(row):
    
    threshold = (row['dob_years']) * 365
    days_employed = row['days_employed']
    
    #пропуски мы возвращаем на место, т.к заполнить их без медианы нельзя.
    if days_employed == np.NaN:
        return days_employed
    
    #Если стаж больше, чем возраст в днях, то стаж указан в часах и мы делим его на 24
    elif days_employed > threshold:
        return days_employed/24
    
    #В остальных случаях стаж указан в днях и мы возвращаем его на место
    else:
        return days_employed

In [16]:
root_df['days_employed'] = root_df.apply(convert_days_employed, axis=1)

#### Рассчет медианы по типу занятости для подстановки вместо nan и null 

In [17]:
#Группировка по типу занятости и рассчет медианы в каждой группе.
groped_by_income_type_df = root_df.groupby('income_type').median()
groped_by_income_type_df

children  days_employed  dob_years  education_id  \
income_type                                                         
безработный           0.5   15267.235531       38.0           0.5   
в декрете             2.0    3296.759962       39.0           1.0   
госслужащий           0.0    2666.912661       40.0           1.0   
компаньон             0.0    1531.798930       39.0           1.0   
пенсионер             0.0   15217.221094       60.0           1.0   
предприниматель       0.0     520.848083       42.5           0.0   
сотрудник             0.0    1568.388775       39.0           1.0   
студент               0.0     578.751554       22.0           0.0   

                 family_status_id  debt   total_income  
income_type                                             
безработный                   0.5   0.5  131339.751676  
в декрете                     0.0   1.0   53829.130729  
госслужащий                   0.0   0.0  150447.935283  
компаньон                     0.0   0.0  172357.950966  
пенсионер                     0.0   0.0  118514.486412  
предприниматель               0.5   0.0  499163.144947  
сотрудник                     0.0   0.0  142594.396847  
студент                       4.0   0.0   98201.625314

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    Медиана пенсионеров все-равно получилась довольно низкой
</div>

<div class="alert alert-success">
<h2> Комментарии от ревьюера №1</h2>
    
Супер, молодец, граммотный подход
</div>

In [18]:
total_income_by_type = {}

for row in groped_by_income_type_df.iterrows():
    
    total_income_by_type[row[0]] = row[1][6]
    
total_income_by_type

{'безработный': 131339.7516762103,
 'в декрете': 53829.13072905995,
 'госслужащий': 150447.9352830068,
 'компаньон': 172357.95096577113,
 'пенсионер': 118514.48641164352,
 'предприниматель': 499163.1449470857,
 'сотрудник': 142594.39684740017,
 'студент': 98201.62531401133}

In [19]:
def fill_na_by_median(row):
    
    #Если параметр days_employed = null, заменяем его на соответсующую медиану из словаря total_income_by_type
    if pd.isna(row['total_income']):
        return total_income_by_type[row['income_type']]
    
    #В остальных случаях возвращаем существующее значение
    else:
        return row['total_income']

In [20]:
root_df['total_income'] = root_df.apply(fill_na_by_median, axis=1)

<div class="alert alert-warning">
<h2> Комментарии от ревьюера №1</h2>
    
Ну тут ты по моему уже усложняешь, можно было сделать код и без словаря
</div>

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>Только что понял, что можно было прям из фрейма с медианами брать)
</div>

#### Решение проблемы с часами вместо дней 

Первое, что нужно сделать - оценить, когда нужно менять значение с часов на дни, а когда оставлять как есть. Для этого определим пороговое значение максимально возможного стажа в днях. 

In [21]:
# Получим словарь в котором ключами является группа, а
# значениями медиана по столбцу days_employed для этой группы

total_days_employed_by_type = {}

for row in groped_by_income_type_df.iterrows():
    total_days_employed_by_type[row[0]] = row[1][1]

total_days_employed_by_type

{'безработный': 15267.235531008522,
 'в декрете': 3296.7599620220594,
 'госслужащий': 2666.912661183607,
 'компаньон': 1531.7989297242334,
 'пенсионер': 15217.221094405466,
 'предприниматель': 520.8480834953765,
 'сотрудник': 1568.3887753757917,
 'студент': 578.7515535382181}

<div class="alert alert-danger">
<h2> Комментарии от ревьюера №1</h2>
    
Обрати внимание какие огромные значения у тебя получились для безработного и пенсионера
    
Там проблема в том, что с определенного возраста стаж учитывается в часах, определи с какого и приведи значения в норму
</div>

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    Про возраст - не уверен, есть записи, в которых возраст 50-52 и указано в часах, а есть где 55+ и указано в днях. Поищу другой критерий.
</div>

In [22]:
def fill_na_by_median(row):
    
    #Если параметр days_employed = null, заменяем его на соответсующую медиану из словаря total_income_by_type
    if pd.isna(row['days_employed']):
        return total_days_employed_by_type[row['income_type']]
    
    #В остальных случаях возвращаем существующее значение
    else:
        return row['days_employed']

In [23]:
root_df['days_employed'] = root_df.apply(fill_na_by_median, axis=1)

#### Проверка на наличие пропусков

In [24]:
root_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

В столбцах ***days_employed*** и ***total_income*** есть пропущенные значения.
1. Первой причниой отсутсвия этих величин может быть связано с тем, что заемщик не устроен официально и не предоставил документы банку. 
2. Второй причиной может являться нарушение в бизнес-процессе, при котором часть данных не попадает в выборку.

Так-как вероятность того, что банк выдаст кредит человеку без подтвержденного дохода крайне низка, то вторая причина является более подходящей.  

Пропущенные значения можно заменить средним, медианой или модой.
1. **Мода** - не подходит, так-как величина з.п. непрерывная, что ведет к очень большому разнообразию.
2. **Среднее** - очень чувствительно к разборосу, что может повлечь его смещение. 
3. **Медиана** - менее остальных метрик чувствительна к разбросу и разнообразию величины

Медиану необходимо подобрать по группе, например, по типу занятности, чтобы получить более реальные значения пропусков.

Пропуски заполнены при помощи рассчета медианы для каждой группы и подставки вместо NaN/Null по соответсвию группе.

### Замена типа данных

In [25]:
root_df['days_employed'] = root_df['days_employed'].astype('int')

In [26]:
root_df['total_income'] = root_df['total_income'].astype('int')

In [27]:
root_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Произведена замена значений в колонках *days_employed* и *total_income* при помощи метода *astype()*
метод замены обусловлен тем, что он короче в написании по сравнению с pd.to_numeric(), а нам не требуется обрабатывать ошибки в данной задаче, т.к. все числа в столбцах имеют тип float и могут быть преобразованы без ошибок.

### Обработка дубликатов

#### Приведение строк в столбцах к нижнему регистру

In [28]:
root_df['education'] = root_df['education'].str.lower()

In [29]:
root_df['income_type'] = root_df['income_type'].str.lower()

In [30]:
root_df['family_status'] = root_df['family_status'].str.lower()

#### Поиск и удаление полных дубликатов

In [31]:
root_df.duplicated().value_counts()

False    21454
True        71
dtype: int64

In [32]:
root_df = root_df.drop_duplicates()

### Вывод

После приведения всех строк к нижнему регистру обнаружено 71 **полных** дубликата. Причиной их возникновения может быть только ошибка сбора данных, так как полное совпадение 12 параметров - практически невозможно. Однако, есть вероятность появление дубликатов в результате замены пропущенных значений на групповые медианы. Использованы методы duplicated() и drop_duplicates() для удаления полных дубликатов, по умолачнию (без передачи параметров) метод находит полные дубликаты (совпадение 12 параметров) и оставляет первый найденный. 

### Лемматизация

In [33]:
from pymystem3 import Mystem
stemmer = Mystem()

In [34]:
root_df.groupby('purpose').sum()

children  days_employed  dob_years  \
purpose                                                                      
автомобили                                   218        2335206      20962   
автомобиль                                   211        2428665      21575   
высшее образование                           207        2003447      19445   
дополнительное образование                   250        2236942      20282   
жилье                                        338        2837369      27334   
заняться высшим образованием                 256        2263137      21426   
заняться образованием                        209        1850288      17835   
на покупку автомобиля                        227        2233578      20423   
на покупку подержанного автомобиля           231        2245567      21026   
на покупку своего автомобиля                 237        2168077      21554   
на проведение свадьбы                        383        3559694      32840   
недвижимость                                 297        2780966      27183   
образование                                  205        2143260      19207   
операции с жильем                            307        2884568      28161   
операции с коммерческой недвижимостью        328        2982384      27858   
операции с недвижимостью                     312        3144077      28936   
операции со своей недвижимостью              319        2971996      27629   
покупка жилой недвижимости                   295        2545373      26325   
покупка жилья                                318        3018548      27411   
покупка жилья для сдачи                      333        2953582      27757   
покупка жилья для семьи                      295        3048909      28270   
покупка коммерческой недвижимости            317        3053561      29012   
покупка недвижимости                         289        2863254      26992   
покупка своего жилья                         316        2689384      26723   
получение высшего образования                184        1865677      18422   
получение дополнительного образования        217        2157536      19642   
получение образования                        232        1953623      18844   
приобретение автомобиля                      222        2282333      19996   
профильное образование                       192        1953696      19073   
ремонт жилью                                 282        2655648      25806   
свадьба                                      368        3591470      34299   
свой автомобиль                              247        2049385      20636   
сделка с автомобилем                         212        2132448      19511   
сделка с подержанным автомобилем             229        2147174      21520   
строительство жилой недвижимости             275        2934375      26949   
строительство недвижимости                   311        2777477      26968   
строительство собственной недвижимости       297        2965049      27420   
сыграть свадьбу                              344        3652614      33089   

                                        education_id  family_status_id  debt  \
purpose                                                                        
автомобили                                       366               522    44   
автомобиль                                       396               454    42   
высшее образование                               361               425    40   
дополнительное образование                       380               480    38   
жилье                                            521               668    46   
заняться высшим образованием                     417               463    43   
заняться образованием                            331               346    39   
на покупку автомобиля                            399               467    44   
на покупку подержанного автомобиля               391               497    36   
на покупку своего автомобиля           

<div class="alert alert-warning">
<h2> Комментарии от ревьюера №1</h2>
    
Зачем группировка? 🤔
</div>

<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    Группировку выполнял вместо counter, визуально по данным выделил группы для категоризации для следующего шага, посидел, подумал и понял, что для большего количества данных такое может уже и не сработать. В будущем так делать не буду)
</div>

In [35]:
def lemmatize_column(row):
    lemma = stemmer.lemmatize(row['purpose'])
    lemma = ''.join(lemma)
    return lemma.rstrip()

In [36]:
root_df['purpose'] = root_df.apply(lemmatize_column, axis=1)

### Вывод

Из всех данных только значение столбца *purpose* требует лемматизации. Это требуется для последующей категоризации данных. 

### Категоризация данных

#### Категоризация по предназначению кредита

In [37]:
def set_purpose_type(row):
    
    if 'автомобиль' in row['purpose']:
        return "Автомобильные кредиты"
    
    elif 'жилье' in row['purpose']:
        return "Жилищные кредиты"
    
    elif 'жилье' in row['purpose']:
        return "Жилищные кредиты"
    
    elif 'образование' in row['purpose']:
        return "Кредиты на образование"
    
    elif 'недвижимость' in row['purpose']:
        return "Кредиты на недвижимость"
    
    elif 'свадьба' in row['purpose']:
        return "Кредиты на свадьбу"

<div class="alert alert-warning">
<h2> Комментарии от ревьюера №1</h2>
    
Не понятно с чего ты делаешь выводы по данным группам для категоризации, попробуй через Counter посчитать повторение каждого слова в лемматизированном столбце
</div>

In [38]:
root_df['purpose_type'] = root_df.apply(set_purpose_type, axis=1)

#### Категоризация по уровню дохода

In [39]:
total_income_mean = root_df['total_income'].mean()
total_income_mean

165319.57229421087

In [40]:
def set_income_type(row):
    
    if total_income_mean < row['total_income']  < total_income_mean * 2:
        return "Доход выше среднего"
    
    if total_income_mean * 2 < row['total_income']:
        return "Сильно выше среднего"
    
    if total_income_mean > row['total_income']  > total_income_mean / 2:
        return "Доход ниже среднего"
    
    if total_income_mean /2 > row['total_income']:
        return "Сильно ниже среднего"
    
    else:
        return "Средний"
    

In [41]:
root_df['income_level_type'] = root_df.apply(set_income_type, axis=1)

<div class="alert alert-danger">
<h2> Комментарии от ревьюера №1</h2>
    
Давай здесь все-таки выделим хотя бы один "словарь" для образования или семейного статуса, как нас просят в задании. После этого мы теоретически сможем вовсе удалить из основной таблицы столбец со строковым обозначением образования (или семейного статуса), что здорово сэкономит место и время обработки. Просто скопируй пару столбцов, например, 'family_status' и 'family_status_id' в новый датафрейм и удали дубликаты, как было в теории в разделе "Классификация по типу"


<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    То есть словари нужны только для того, чтобы снизить размерность основной таблицы ?
    Словарь создал, соответсвтующий столбец из основного удалил.
</div>

In [42]:
df_dict_education = pd.DataFrame()
df_dict_education['education'] = root_df['education']
df_dict_education['education_id'] = root_df['education_id']
df_dict_education = df_dict_education.drop_duplicates().reset_index(drop=True)
df_dict_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [43]:
root_df = root_df.drop('education',axis='columns')
root_df.head(10)

children  days_employed  dob_years  education_id     family_status  \
0         1           8437         42             0   женат / замужем   
1         1           4024         36             1   женат / замужем   
2         0           5623         33             1   женат / замужем   
3         3           4124         32             1   женат / замужем   
4         0          14177         53             1  гражданский брак   
5         0            926         27             0  гражданский брак   
6         0           2879         43             0   женат / замужем   
7         0            152         50             1   женат / замужем   
8         2           6929         35             0  гражданский брак   
9         0           2188         41             1   женат / замужем   

   family_status_id gender income_type  debt  total_income  \
0                 0      F   сотрудник     0        253875   
1                 0      F   сотрудник     0        112080   
2                 0      M   сотрудник     0        145885   
3                 0      M   сотрудник     0        267628   
4                 1      F   пенсионер     0        158616   
5                 1      M   компаньон     0        255763   
6                 0      F   компаньон     0        240525   
7                 0      M   сотрудник     0        135823   
8                 1      F   сотрудник     0         95856   
9                 0      M   сотрудник     0        144425   

                      purpose            purpose_type    income_level_type  
0               покупка жилье        Жилищные кредиты  Доход выше среднего  
1     приобретение автомобиль   Автомобильные кредиты  Доход ниже среднего  
2               покупка жилье        Жилищные кредиты  Доход ниже среднего  
3  дополнительный образование  Кредиты на образование  Доход выше среднего  
4             сыграть свадьба      Кредиты на свадьбу  Доход ниже среднего  
5               покупка жилье        Жилищные кредиты  Доход выше среднего  
6            операция с жилье        Жилищные кредиты  Доход выше среднего  
7                 образование  Кредиты на образование  Доход ниже среднего  
8       на проведение свадьба      Кредиты на свадьбу  Доход ниже среднего  
9     покупка жилье для семья        Жилищные кредиты  Доход ниже среднего

### Вывод

Так-как по заданию необходимо ответить на три вопроса, то категоризацию данных необходимо провести таким образом, чтобы полученные категорию наблюдений помогли в этом. 

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

##### Группируем по количеству детей и находим количество наблюдений в каждой группе

In [44]:
grouped_by_children = root_df.groupby('children').sum()
grouped_by_children['total_obs'] = root_df.groupby('children')['children'].count()

##### Найдем отношение количества должников к количеству наблюдений, сгруппированных по количеству детей

In [45]:
grouped_by_children['children_to_debt'] = grouped_by_children.debt/grouped_by_children.total_obs*100
grouped_by_children['children_to_debt'].sort_values(ascending=False)

children
4    9.756098
2    9.492481
1    9.165808
3    8.181818
0    7.543822
5    0.000000
Name: children_to_debt, dtype: float64

### Вывод

Чаще всего не отдают кредиты люди, у которых в семье 1, 2 или 4 ребенка, это случается *9.1-9.8%* случаев.
Люди, у которых 3 ребенка или нет детей совсем отдают кредиты в срок чаще, *8.2%* и *7.5%* соответсвенно.
Люди, у которых 5 детей - отдают кредиты в 100% случаев, что очень похоже на ошибку при сборе данных, однако, заменить эти значения не представляется возможным.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

##### Группируем по семейному положению и находим количество наблюдений в каждой группе

In [46]:
grouped_by_family_status = root_df.groupby('family_status').sum()
grouped_by_family_status['total_obs'] = root_df.groupby('family_status')['family_status'].count()

##### Найдем отношение количества должников к общему количеству наблюдений, сгруппированных по семейному положению

In [47]:
grouped_by_family_status['family_status_to_debt'] = grouped_by_family_status.debt/grouped_by_family_status.total_obs*100
grouped_by_family_status['family_status_to_debt'].sort_values(ascending=False)

family_status
не женат / не замужем    9.750890
гражданский брак         9.347145
женат / замужем          7.545182
в разводе                7.112971
вдовец / вдова           6.569343
Name: family_status_to_debt, dtype: float64

### Вывод

Люди **не состоящие в браке** или **состоящие в гражданском браке** - чаще всего не возвращают кредиты, *9.75%* и *9.35%* соответсвенно. **Разведенные люди** и люди, **состоящие в официальном браке** не возвращают кредиты в *7.55%* и *7.11%* случаях соответсвенно. **Вдовы и вдовцы** наименее рискованная категория с *6.6%*

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

##### Группируем по уровню дохода и находим количество наблюдений в каждой группе

In [48]:
grouped_by_income_level_type = root_df.groupby('income_level_type').sum()
grouped_by_income_level_type['total_obs'] = root_df.groupby('income_level_type')['income_level_type'].count()
grouped_by_income_level_type

children  days_employed  dob_years  education_id  \
income_level_type                                                        
Доход выше среднего       3468       26914842     299097          5322   
Доход ниже среднего       5192       50487029     464590          9199   
Сильно выше среднего       591        4007692      46550           569   
Сильно ниже среднего      1059       16950774     118104          2440   

                      family_status_id  debt  total_income  total_obs  
income_level_type                                                      
Доход выше среднего               6783   546    1544342632       7080  
Доход ниже среднего              10773   920    1343137330      10724  
Сильно выше среднего              1006    78     492760871       1091  
Сильно ниже среднего              2332   197     166525271       2559

##### Найдем отношение количества должников к общему количеству наблюдений, сгруппированных по уровню дохода

In [49]:
grouped_by_income_level_type['income_level_to_debt']=(
    grouped_by_income_level_type.debt/grouped_by_income_level_type.total_obs*100
)

In [50]:
grouped_by_income_level_type['income_level_to_debt'].sort_values(ascending=False)

income_level_type
Доход ниже среднего     8.578888
Доход выше среднего     7.711864
Сильно ниже среднего    7.698320
Сильно выше среднего    7.149404
Name: income_level_to_debt, dtype: float64

### Вывод

Люди, у которых доход ниже среднего чаще остальных не отдают кредиты. При этом люди, у которых доход сильно ниже среднего отдают кредиты чаще, возможно, это связанно с льготными условиями или с тем фактом, чтом им готовы одобрить лишь незначительные суммы. 

- Как разные цели кредита влияют на его возврат в срок?

##### Группируем по назначению кредита и находим количество наблюдений в каждой группе

In [51]:
grouped_by_purpose_type = root_df.groupby('purpose_type').sum()
grouped_by_purpose_type['total_obs'] = root_df.groupby('purpose_type')['purpose_type'].count()

##### Найдем отношение количества должников к общему количеству наблюдений, сгруппированным по назначению кредита

In [52]:
grouped_by_purpose_type['purpose_to_debt'] = grouped_by_purpose_type.debt/grouped_by_purpose_type.total_obs*100
grouped_by_purpose_type['purpose_to_debt'].sort_values(ascending=False)

purpose_type
Автомобильные кредиты      9.359034
Кредиты на образование     9.220035
Кредиты на свадьбу         8.003442
Кредиты на недвижимость    7.463392
Жилищные кредиты           6.905830
Name: purpose_to_debt, dtype: float64

### Вывод

Чаще всего имеют задолженности по кредиту люди, которые берут их на **образовательные цели** или на различные **операции с автомобилями**, *9.22%* и *9.33%* соответсвенно. **Жилищные кредиты** возвращаются чаще всего, среди клиентов банка только *6.9%* людей имеющих задолженность по кредиту на операции с жильем. В середине топ по риску находятся **кредиты на недвижимость** *7.4%* и **кредиты на свадьбу** 8%

### Шаг 4. Общий вывод

Мы получили метрики, которые показывают, в каком проценте случаев та или иная категория клиентов банка имеет задолженности по кредиту. Это поможет сформировать примерны портрет потребителя, которому кредит выдавать не стоит. Так же данные очищены и на их основе можно построить модель, решающую задачу регресии, чтобы узнать, насколько сильно каждый фактор влияет на вероятность выплаты кредиты в срок. 

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<div class="alert alert-success">
<h2> Комментарии от ревьюера №1</h2>
    
Ты большой молодец, твой проект выполнен сейчас на 90% и я не могу не отправить тебе его обратно, чтобы мы все таки пришли к 100%, думаю это не займет у тебя много времени.
    
Обрати внимание на следующее:

    1) Предобработка данных (момент со стажем работы)
    2) Категоризация данных
    
Жду скорее твою работу на повторную проверку!


</div>


<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ ОТ УЧЕНИКА:</b>
    Большое спасибо за советы, желтые карточки утчту на будущее. Красные поправил.
</div>